### PR TITLE
New Feature: Caching

### PR DESCRIPTION
**Pull request by 21christiansc**

### CODE REVIEW

File: src/Storage/AssetMapCache.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	 */
+	public function getMapCache() : ?AssetMap
+	{
+		$cacheItem = $this->cachePool->getItem(self::CACHE_KEY);
+
+		if($cacheItem->isHit()) {
+			return $cacheItem->get();
+		}
+
+		return null;
```
Comment (apfelbox) [inline src/Storage/AssetMapCache.php:57]:
```suggestion
	return $cacheItem->isHit()
		? $cacheItem->get()
		: null;
```

File: src/Storage/AssetMapCache.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Component\Console\Input\ArrayInput;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Console\Output\NullOutput;
+use Symfony\Component\HttpKernel\KernelInterface;
+use Torr\Assets\Command\AssetsDumpCommand;
+
+final class AssetMapCache
+{
+	private const CACHE_KEY = '21torr.asset.asset_map';
```
Comment (apfelbox) [inline src/Storage/AssetMapCache.php:14]:
```suggestion
	private const CACHE_KEY = '21torr.assets.asset_map';
```

File: src/Storage/AssetMapCache.php [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		$this->kernel = $kernel;
+		$this->storage = $storage;
+	}
+
+	/**
+	 * save AssetMap in Cache
+	 */
+	public function setMapCache(AssetMap $map) : void
+	{
+		$cacheItem = $this->cachePool->getItem(self::CACHE_KEY);
```
Comment (apfelbox) [inline src/Storage/AssetMapCache.php:41]: Should we store the item in a property?

File: src/Storage/AssetMapCache.php [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	 */
+	public function clearCache() : void
+	{
+		$this->cachePool->deleteItem(self::CACHE_KEY);
+	}
+
+	/**
+	 * execute AssetsDumpCommand
+	 */
+	private function executeDumpCommand() : void
```
Comment (apfelbox) [inline src/Storage/AssetMapCache.php:93]: You should basically never call CLI commands in your application. Just fetch the service the CLI command uses internally and call it directly

File: src/Storage/AssetMapCache.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		return null;
+	}
+
+	/**
+	 * get AssetMap
+	 * in debug mode return new empty AssetMap and clear dump directory
+	 */
+	public function getAssetMap() : ?AssetMap
+	{
+		if($this->kernel->isDebug() === true) {
```
Comment (apfelbox) [inline src/Storage/AssetMapCache.php:66]:
```suggestion
		if ($this->kernel->isDebug()) {
```
An `if` already expects a `bool`, so no need to transform a `bool` to a `bool`.

File: src/Storage/AssetMapCache.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		$application->setAutoExit(false);
+
+		$input = new ArrayInput([
+			'command' => AssetsDumpCommand::getDefaultName()
+		]);
+		$output = new NullOutput();
+
+		$application->run($input, $output);
+	}
+}
```
Comment (apfelbox) [inline src/Storage/AssetMapCache.php:105]: Please add a newline at the end of the file

File: src/Storage/AssetMapCache.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	/**
+	 * get AssetMap
+	 * in debug mode return new empty AssetMap and clear dump directory
+	 */
+	public function getAssetMap() : ?AssetMap
+	{
+		if($this->kernel->isDebug() === true) {
+			$this->storage->clearStorage();
```
Comment (apfelbox) [inline src/Storage/AssetMapCache.php:67]: It looks weird, that the asset map cache is clearing the storage.

We should add a top level service, something like an `AssetManager` that is the top-level component to get the asset map, dump the assets, handle the cache, etc...

**Review by apfelbox — CHANGES_REQUESTED** [review]

File: src/Storage/AssetMapCache.php (hunk 1 above)
Comment (21christiansc) [reply src/Storage/AssetMapCache.php:57]: Changes are made in a new branch

File: src/Storage/AssetMapCache.php (hunk 2 above)
Comment (21christiansc) [reply src/Storage/AssetMapCache.php:14]: Changes are made in a new branch

File: src/Storage/AssetMapCache.php (hunk 4 above)
Comment (21christiansc) [reply src/Storage/AssetMapCache.php:93]: Changes are made in a new branch

File: src/Storage/AssetMapCache.php (hunk 5 above)
Comment (21christiansc) [reply src/Storage/AssetMapCache.php:66]: Changes are made in a new branch

File: src/Storage/AssetMapCache.php (hunk 6 above)
Comment (21christiansc) [reply src/Storage/AssetMapCache.php:105]: Changes are made in a new branch

File: src/Storage/AssetMapCache.php (hunk 7 above)
Comment (21christiansc) [reply src/Storage/AssetMapCache.php:67]: Changes are made in a new branch

File: src/Storage/AssetMapCache.php (hunk 3 above)
Comment (21christiansc) [reply src/Storage/AssetMapCache.php:41]: File was changed in a new branch

**Comment by apfelbox** [comment]
We should really fix the CI in the next PRs

**Pull request merged**